src/rep.c

```diff
@@ -0,0 +1,210 @@
+#include "vctrs.h"
+#include "utils.h"
+
+// Initialised at load time
+static struct vctrs_arg args_times_;
+static struct vctrs_arg args_x_;
+static struct vctrs_arg* const args_times = &args_times_;
+static struct vctrs_arg* const args_x = &args_x_;
+
+static inline void stop_rep_size_oob(long long size);
+static inline void stop_rep_times_size();
+static inline void stop_rep_times_negative();
+static inline void stop_rep_times_missing();
+static inline void stop_rep_each_times_negative(R_len_t i);
+static inline void stop_rep_each_times_missing(R_len_t i);
+
+// -----------------------------------------------------------------------------
+
+static SEXP vec_rep(SEXP x, R_len_t times) {
+  if (times < 0) {
+    if (times == NA_INTEGER) {
+      stop_rep_times_missing();
+    } else {
+      stop_rep_times_negative();
+    }
+  }
+
+  const R_len_t x_size = vec_size(x);
+
+  const long long temp_size = (long long) x_size * times;
```

Comment:
Could you point me to the overflow check that you would use here?

Reply:
I've never done such a check so I'd use the first SO result: https://stackoverflow.com/questions/1815367/catch-and-compute-overflow-during-multiplication-of-two-large-integers

Reply:
The problem with the first one listed there is that it detects overflow _after_ it happens. On USBAN this will throw an overflow warning if we have any tests that try and check the behavior when this happens. Signed integer overflow also has undefined behavior, so I don't think we can rely on this working everywhere.

```
x = a * b;
if (a != 0 && x / a != b) {
    // overflow handling
}
```

Finding it _before_ it happens for multiplication seems to be quite hard (even R doesn't do it). It is possible for addition/subtraction, and division doesn't require it.

This is R not doing it for multiplication, with a note about USBAN:
https://github.com/wch/r-source/blob/dcf859833bb344a5f5e132ff64b76b7175a2ed5d/src/main/arithmetic.c#L372

Here they check if before it happens in addition
https://github.com/wch/r-source/blob/dcf859833bb344a5f5e132ff64b76b7175a2ed5d/src/main/arithmetic.c#L343

The only alternative option I've seen is the "computing the carry" one also listed in that SO post. It seems quite involved though.

Reply:
Undefined behaviour is not prohibited behaviour, it's probably ok to rely on it in this case. We'd need to avoid having it covered by unit test so UBSAN doesn't warn about it.

The second answer looks more promising though?

Reply:
FYI this SO answer is all about unsigned int64, so that might make it less useful

Reply:
gcc advises to cast to unsigned before checking for overflow: https://www.gnu.org/software/autoconf/manual/autoconf-2.63/html_node/Signed-Overflow-Advice.html

Reply:
Seems helpful: https://codereview.stackexchange.com/questions/98791/safe-multiplication-of-two-64-bit-signed-integers

Reply:
Okay, I've summarized that last post (plus one other I found) with a proof and implementation here.
https://github.com/DavisVaughan/int64/issues/1#issuecomment-604052348

We only care about positive * positive here, so I'll simplify the implementation for `vec_rep`

Reply:
Okay so I think that this int64_t overflow detection will be useful elsewhere, but I have now come full circle on this.

I think that the best thing to do here (keeping long vector support in mind) is to coerce to double, multiply, and check if the result is less than `R_LEN_T_MAX`. When we support long vectors:

- The `times` vector will be provided by the user as a `double`, so preemptively using double multiplication makes sense.

- We will just have to change `R_LEN_T_MAX` for `R_XLEN_T_MAX` and a few `R_len_t` to `R_xlen_t`

Reply:
Mind blown! I didn't know floating point computations handled overflow elegantly, that's good to know.

Your idea of using double makes a lot of sense to me.